default LBP: Warn on misconfigured local DC

wprzytula opened the pull request:
Ref: #1371

## What's done
`DefaultPolicy` now issues respective warnings:
1. if the preferred datacenter is not present in the cluster;
2. if all nodes in the preferred datacenter are disabled by the `HostFilter`.

This helps avoid confusion when the user expects requests to be sent to a specific datacenter, but it is not available.

### Note
As warnings are going to be emitted on every request (because the configuration is most likely session-wide), the messages will quickly flood the logs and the user should notice.

Alternatively, we could implement some frequency limiting of those warnings, but we prefer to:
- keep it simple for now,
- let the user notice the misconfiguration as quickly as possible.

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [ ] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- ~~[ ] I have provided docstrings for the public items that I want to introduce.~~
- ~~[ ] I have adjusted the documentation in `./docs/source/`.~~
- ~~[ ] I added appropriate `Fixes:` annotations to PR description.~~

wprzytula committed:
default LBP: extract method with sanity checks

`DefaultPolicy::pick()` begins with some sanity checks, whose goal is
to warn the user about misconfiguration. Those checks are extracted
into a separate method, `DefaultPolicy::pick_sanity_checks()`, so that
`pick()` is decluttered. The next commits introduce more sanity checks.

wprzytula committed:
default LBP: pick_sanity_checks() slight refactor

This makes the next commit cleaner.

wprzytula committed:
default LBP: warn on preferred DC misconfiguration

DefaultPolicy now issues respective warnings:
1. if the preferred datacenter is not present in the cluster;
2. if all nodes in the preferred datacenter are disabled by
   the HostFilter.

This helps avoid confusion when the user expects requests to be sent
to a specific datacenter, but it is not available.

As warnings are going to be emitted on every request (because
the configuration is most likely session-wide), the messages will
quickly flood the logs and the user should notice.
Alternatively, we could implement some frequency limiting of those
warnings, but we rather prefer to:
- keep it simple for now,
- let the user notice the misconfiguration as quickly as possible.

Lorak-mmk commented on scylla/src/policies/load_balancing/default.rs:
```diff
@@ -602,6 +588,77 @@ impl DefaultPolicy {
+                }
+            }
+
+            // Verify that the preferred datacenter is actually present in the cluster.
+            // If not, shout a warning.
+            if !cluster.all_nodes.iter().any(|node| {
+                node.datacenter
+                    .as_deref()
+                    .is_some_and(|dc| dc == preferred_dc)
+            }) {
```
This iterates over potentially all nodes (even in a happy case). I don't think it is a good idea performance-wise to have such checks in the hot path.

Lorak-mmk replied:
Can't you use ReplicaLocator for that? I see it has `ReplicationInfo` which has:
```rust
    /// We keep a separate token ring for each datacenter.
    /// Each datacenter's ring contains only nodes from this datacenter.
    /// It makes it simpler and more efficient to calculate the NetworkTopologyStrategy replicas.
    datacenters: HashMap<String, DatacenterNodes>,
```

wprzytula replied:
Good catch!

Lorak-mmk commented on scylla/src/policies/load_balancing/default.rs:
```diff
@@ -602,6 +588,77 @@ impl DefaultPolicy {
+You won't be able to execute any requests!"
+                    );
+                }
+            }
+            // Verify that there exist any enabled nodes in the preferred datacenter.
+            // If not, shout a warning.
+            else if !cluster.all_nodes.iter().any(|node| {
+                node.is_enabled()
+                    && node
+                        .datacenter
+                        .as_deref()
```
ditto

wprzytula committed:
observability: Introduce RateLimiter & macros

RateLimiter is a new utility that allows rate-limiting arbitrary
actions. It provides one central method, `try_acquire`, which checks if
the action should be performed based on a specified interval. If the
action has not been performed:
1. ever, or
2. in the last `interval`,
it updates the last-performed timestamp and returns `true`, allowing the
action to proceed. Otherwise, it returns `false`, indicating the action
should be skipped due to rate limiting.

This utility is designed to be efficient and thread-safe, using atomic
operations to ensure that multiple threads can safely check and update
the last-performed timestamp without locking.

The `RateLimiter` is intended for use in scenarios where actions need to
be performed at a controlled rate, such as logging, warning messages,
or other operations that should not be executed too frequently to avoid
spamming logs or overwhelming resources.

For now, it's used to rate-limit warning messages emitted by
the driver, in the implementation of `warn_rate_limited!` macro.

NOTE to reviewers: this code has been generated with major help of
the Claude Sonnet 4 AI model. Beware of common pitfalls of AI-generated
code.

wprzytula committed:
default LBP: rate limit sanity checks warnings

In case of misconfiguration, the default load balancing policy can throw
a lot of warnings, which flood the logs and overwhelm the machine.
To prevent this, we rate limit the warnings there to one per second.

NOTE: each warning is rate limited separately, so if there are multiple
misconfigurations, each will still be logged.

wprzytula committed:
LBPs: rate limit various warnings

There were some more warnings that were not rate-limited, which could
flood the logs in case of misconfiguration. This commit adds rate
limiting there, too.

wprzytula commented:
Rebased on main.

Lorak-mmk reviewed (changes requested):
I think the API of the newly introduced RateLimiter could be better. My idea:
- Duration should be stored inside the limiter, because I find it unlikely to that we'll need to use multiple durations with single limiter.
- Instead of having a method returning bool, we can have `maybe_execute` taking a closure that will be called if timer passed. 
- After the above, the macros won't really give much value over having the limiter stored explicitly inside DefaultPolicy (or having it explicitly as a static). Let's avoid macros unless they really do help in something.
- Can you also move timestamp generator to use the new rate limiter?

Lorak-mmk commented:
> The idea I had with the macros there was to make rate-limited warnings maximally welcoming, so that there's minimum effort and code bloat involved when employing them. Keep in mind that the way macro encapsulates the static instance of the RateLimiter makes it immune to accidental sharing of the same limiter for two separate warnings.

In case of DefaultPolicy I don't see any reason to have a separate limiters for those 4 warnings. Each limiter is another global object that is always present. Macros also hide this overhead.

wprzytula commented:
> > This is limiting. This will, for instance, disallow for executing async code.
> 
> This can be solved by having also an async version of this method. What are the other limitations?

WDYM by the async version?
To have one method accepting a closure and one accepting an `impl Future`?

This seems bloaty. Why not just give users a flexible API that lets them utilize the `RateLimiter` as they want?

wprzytula commented:
> Each limiter is another global object that is always present. Macros also hide this overhead.

This overhead is justified: each warning string is, similarly, another static object that is always present. As the (I assume you mean the memory) overhead is tiny and static, I believe it's not anything to bikeshed about.

Lorak-mmk commented:
> This seems bloaty. Why not just give users a flexible API that lets them utilize the RateLimiter as they want?

Because it introduces macros for something that can be done without them, without much more verbosity.

Lorak-mmk commented:
> These warnings are different. We do want the user to get each of them, not only one. If we rate limit them together, then if multiple of them are triggered in close time, only the first one is printed.

It is extremely unlikely for multiple of those warnings to be triggered, I don't see much sense in caring about this case.

dkropachev commented:
> > Each limiter is another global object that is always present. Macros also hide this overhead.
> 
> This overhead is justified: each warning string is, similarly, another static object that is always present. As the (I assume you mean the memory) overhead is tiny and static, I believe it's not anything to bikeshed about.

What bothers me is that on every request driver runs these checks, which is not realy necessary.
What if we would do that only when driver get into undesired state, say empty query plan or failover to another DC ?

wprzytula commented:
> > This seems bloaty. Why not just give users a flexible API that lets them utilize the RateLimiter as they want?
> 
> Because it introduces macros for something that can be done without them, without much more verbosity.

You seem to mix two things. I'm arguing about exposing a boolean from the `RateLimiter`'s method, in contrast to only accepting a closure/Future. You are arguing that macros are unneeded. These are two distinct things.

Lorak-mmk commented:
> I can agree with this reasoning, but I still believe that there's no generic enough way to represent "an action" to create such API with just one method.

I really don't see the problem with having 2 - try_execute and try_execute_async.
Rust doesn't have effect system / keyword generics, and afaik this is the most typical way to deal with having sync / async version of something.

wprzytula commented:
> > > Each limiter is another global object that is always present. Macros also hide this overhead.
> > 
> > 
> > This overhead is justified: each warning string is, similarly, another static object that is always present. As the (I assume you mean the memory) overhead is tiny and static, I believe it's not anything to bikeshed about.
> 
> What concerns me is that the driver performs these checks on every request, even though they’re usually unnecessary. In effect, we’re sacrificing performance in the common case just to show a helpful warning when the user misconfigures the driver. What if we would do that only when driver get into undesired state, say empty query plan or failover to another DC ?

I understand your concerns.

However, if think these checks should be cheap. @Lorak-mmk do you agree?
-
